common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -8577,6 +8577,38 @@ static void NativeCrypto_SSL_free(JNIEnv* env, jclass, jlong ssl_address, CONSCR
     SSL_free(ssl);
 }
 
+static jbyteArray get_session_id(JNIEnv* env, SSL_SESSION* ssl_session) {
+    unsigned int length;
+    const uint8_t* id = SSL_SESSION_get_id(ssl_session, &length);
+    JNI_TRACE("ssl_session=%p get_session_id id=%p length=%u", ssl_session, id, length);
+    if (id && length > 0) {
+        jbyteArray result = env->NewByteArray(static_cast<jsize>(length));
+        if (result != nullptr) {
+            const jbyte* src = reinterpret_cast<const jbyte*>(id);
+            env->SetByteArrayRegion(result, 0, static_cast<jsize>(length), src);
+        }
+        return result;
+    }
+    if (SSL_SESSION_has_ticket(ssl_session)) {
+        JNI_TRACE("ssl_session=%p get_session_id has_ticket=true", ssl_session);
+        const uint8_t* ticket;
+        size_t tick_len;
+        SSL_SESSION_get0_ticket(ssl_session, &ticket, &tick_len);
+        uint8_t data[32];
+        if (!EVP_Digest(ticket, tick_len, data, &length, EVP_sha256(), NULL)) {
```

Comment:
I'm OOO this week and haven't looked at the rest of this, but this part looks wrong.

I take it Conscrypt relies on OpenSSL's weird behavior of faking a session ID for ID-less sessions (which already exist in TLS 1.2)? This is probably indicative of a bug in Conscrypt's session-handling. (E.g. the other PR revealed problems with the client session cache keying things wrong.) In that case, we should fix that.

If not and Conscrypt legitimately wants this field filled in, we should fix BoringSSL to consistently do this in TLS 1.3 as well, rather than add hacks to Conscrypt.

@dvorak42 FYI

Reply:
Yeah, there's a standard Java API ([`SSLSessionContext`](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLSessionContext.html)) that's specified in terms of IDs, and to continue to support it we want to have IDs for sessions in the cache.  I'm happy enough to have this done in BoringSSL instead, but I didn't want to assume you wanted that.

(FYI, we can chat about this over e-mail and/or when you're back if that's easier, I was planning to wait until you could take a good look at this before submitting.)